File: client/lib/models/responses/post.dart
/// One feed post. Mirrors backend `PostDto`. The `payload` JSONB snapshot
/// is the per-type render data captured at publish time — the card reads
/// from this map without dereferencing the soft link to the source row.
class Post {
  Post({
    required this.id,
    required this.authorId,
    required this.type,
    required this.visibility,
    required this.likeCount,
    required this.commentCount,
    required this.createdAt,
    this.title,
    this.note,
    this.sourceRefType,
    this.sourceRefId,
    this.payload,
  });

  final int id;
  final int authorId;

  /// One of {@code workout | run | cycle | evolution | manual}.
  final String type;
  final String? title;
  final String? note;
  final String? sourceRefType;
  final int? sourceRefId;

  /// Per-type snapshot: workout has `totalVolumeKg`/`totalSets`/`prCount`,
  /// run/cycle has `distanceM`/`durationSeconds`/`avgHr?`/`kcal?`,
  /// evolution has `weightKg`/`bodyFatPct?`/`bfMethod?`, manual is null.
  final Map<String, dynamic>? payload;

  /// One of {@code public | followers | private}.
  final String visibility;
  final int likeCount;
  final int commentCount;
  final DateTime createdAt;

  factory Post.fromJson(Map<String, dynamic> json) {
    return Post(
      id: (json['id'] as num).toInt(),
      authorId: (json['authorId'] as num).toInt(),
      type: json['type'] as String,
      title: json['title'] as String?,
      note: json['note'] as String?,
      sourceRefType: json['sourceRefType'] as String?,
      sourceRefId: (json['sourceRefId'] as num?)?.toInt(),
      payload: json['payload'] == null
          ? null
          : Map<String, dynamic>.from(json['payload'] as Map),
      visibility: json['visibility'] as String? ?? 'followers',
      likeCount: (json['likeCount'] as num?)?.toInt() ?? 0,
      commentCount: (json['commentCount'] as num?)?.toInt() ?? 0,
      createdAt: DateTime.parse(json['createdAt'] as String),
    );
  }

  /// Returns a copy with mutable fields swapped — used by the feed card
  /// for optimistic like-count flips that get reconciled on the API
  /// response.
  Post copyWith({int? likeCount, int? commentCount}) {
    return Post(
      id: id,
      authorId: authorId,
      type: type,
      title: title,
      note: note,
      sourceRefType: sourceRefType,
      sourceRefId: sourceRefId,
      payload: payload,
      visibility: visibility,
      likeCount: likeCount ?? this.likeCount,
      commentCount: commentCount ?? this.commentCount,
      createdAt: createdAt,
    );
  }
}

File: api/src/main/java/com/example/athletehub/model/Post.java
package com.example.athletehub.model;

import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.JdbcTypeCode;
import org.hibernate.type.SqlTypes;

import java.time.OffsetDateTime;
import java.util.Map;

/**
 * One row per published feed item. {@code type} discriminates how the
 * card renders ({@code workout | run | cycle | evolution | manual}).
 * {@code sourceRefType} + {@code sourceRefId} are a soft link back to
 * the row that triggered the auto-post — no FK because we don't want a
 * source-row delete to be blocked by a post; the link goes stale
 * gracefully.
 *
 * <p>{@code payload} is a JSONB snapshot of what the card rendered at
 * publish time (stats, sparkline values, before/after labels). Snapshot
 * semantics mean a coach renaming an exercise tomorrow doesn't rewrite
 * yesterday's feed cards. {@code @JdbcTypeCode(SqlTypes.JSON)} lets us
 * map to a {@code Map<String, Object>} directly — Hibernate handles
 * Jackson serialization on read/write.
 *
 * <p>{@code likeCount} + {@code commentCount} are denormalized counters
 * maintained by the service so a feed card render is O(1). Soft-delete
 * via {@code deletedAt} so threads + counter history stay consistent.
 */
@Entity
@Table(name = "posts")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Post {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "author_id", nullable = false)
    private Long authorId;

    @Column(nullable = false)
    private String type;

    private String title;

    private String note;

    @Column(name = "source_ref_type")
    private String sourceRefType;

    @Column(name = "source_ref_id")
    private Long sourceRefId;

    @JdbcTypeCode(SqlTypes.JSON)
    @Column(columnDefinition = "jsonb")
    private Map<String, Object> payload;

    @Column(name = "image_media_id")
    private Long imageMediaId;

    @Column(nullable = false)
    private String visibility;

    @Column(name = "like_count", nullable = false)
    private int likeCount;

    @Column(name = "comment_count", nullable = false)
    private int commentCount;

    @Column(name = "deleted_at")
    private OffsetDateTime deletedAt;

    @Column(name = "created_at", nullable = false, updatable = false)
    private OffsetDateTime createdAt;

    @PrePersist
    void onCreate() {
        OffsetDateTime now = OffsetDateTime.now();
        if (createdAt == null) createdAt = now;
        if (visibility == null) visibility = "followers";
    }

    public boolean isDeleted() {
        return deletedAt != null;
    }
}
